### PR TITLE
Fix/vfio crashes

### PR DESCRIPTION
**Pull request by n-eiling**
combine functionalities of binaries into a single one

combine what was previously achieved by the separate binaries
villas-fpga-xbar-select and villas-fpga-cat into a single new
binary villas-fpga-ctl. Here we can select crossbar connections
via command line parameters. To avoid regression there are shell
scripts providing the old functionalities directly.

Currently the villas-fpga-pipe functionality is not supported,
because we still need to implement stdin input and routing that
to the fpga.

make DMA ip unmap memory owned by itself

unmapping of the scatter gather attribute memory was done
after the DMA destructor was called, leading to Card trying to
unmap memory that was already freed.
This lead to crashing during cleaning up.

### CODE REVIEW

**Comment by n-eiling** [comment]
This is an alternative to all of the shared_ptr stuff I tried before. I'd like to keep the unique_ptr MemoryBlocks for now.

File: src/villas-fpga-xbar-select.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+CWD=$(dirname -- "$0")
+
+# currently connects RTDS and OPAL-RT
+${CWD}/villas-fpga-cat -c ${CWD}/../../etc/fpgas.json --connect "0<->1" --no-dma
```
Comment (stv0g) [inline src/villas-fpga-xbar-select.sh:19]: `villas-fpga-ctrl`?

File: src/villas-fpga-xbar-select.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+CWD=$(dirname -- "$0")
+
+# currently connects RTDS and OPAL-RT
```
Comment (stv0g) [inline src/villas-fpga-xbar-select.sh:18]: Please start comments with upper case letter :)

File: src/villas-fpga-ctrl.cpp [hunk 3]
```diff
@@ -105,50 +114,48 @@ int main(int argc, char* argv[])
...
-				//logger->info("Read value: {}", dval);
+			// Setup read transfer
+			dma->read(block[0], block[0].getSize());
+			size_t cur = 0, next = 1;
+			while (true) {
+				dma->read(block[next], block[next].getSize());
+				auto bytesRead = dma->readComplete();
+				// Setup read transfer
+
+				//auto valuesRead = bytesRead / sizeof(int32_t);
```
Comment (stv0g) [inline src/villas-fpga-ctrl.cpp:135]: Please remove dead code

File: src/villas-fpga-ctrl.cpp [hunk 4]
```diff
@@ -105,50 +114,48 @@ int main(int argc, char* argv[])
...
+						(int64_t)(mem[1] & 0xFFFF0000) << 16 |
+						(int64_t)(mem[0] & 0xFFFF) << 16 |
+						(int64_t)(mem[0] & 0xFFFF0000) >> 16;
+					double dval = *((double*)(&ival));
+					std::cerr << std::hex << ival << "," << dval << std::endl;
+					bytesRead -= 8;*/
+					//logger->info("Read value: {}", dval);
+				}
+				cur = next;
+				next = (next + 1) % (sizeof(mem)/sizeof(mem[0]));
```
Comment (stv0g) [inline src/villas-fpga-ctrl.cpp:157]: Missing spaces around `/`

File: src/villas-fpga-ctrl.cpp [hunk 5]
```diff
@@ -105,50 +114,48 @@ int main(int argc, char* argv[])
...
+					//std::cerr << std::hex << ival << ",";
+					std::cerr << fval << std::endl;
+					/*int64_t ival = (int64_t)(mem[1] & 0xFFFF) << 48 |
+						(int64_t)(mem[1] & 0xFFFF0000) << 16 |
+						(int64_t)(mem[0] & 0xFFFF) << 16 |
+						(int64_t)(mem[0] & 0xFFFF0000) >> 16;
+					double dval = *((double*)(&ival));
+					std::cerr << std::hex << ival << "," << dval << std::endl;
+					bytesRead -= 8;*/
+					//logger->info("Read value: {}", dval);
```
Comment (stv0g) [inline src/villas-fpga-ctrl.cpp:154]: Dead code

File: src/villas-fpga-ctrl.cpp [hunk 6]
```diff
@@ -48,6 +50,8 @@ using namespace villas;
 static std::shared_ptr<kernel::pci::DeviceList> pciDevices;
 static auto logger = villas::logging.get("cat");
 
+
```
Comment (stv0g) [inline src/villas-fpga-ctrl.cpp:53]: Two empty lines should be sufficient

File: src/villas-fpga-cat.sh [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+CWD=$(dirname -- "$0")
+
+${CWD}/villas-fpga-cat -c ${CWD}/../../etc/fpgas.json --connect "2<->stdout"
```
Comment (stv0g) [inline src/villas-fpga-cat.sh:18]: `villas-fpga-ctrl`

File: src/villas-fpga-cat.sh [hunk 8]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (stv0g) [inline src/villas-fpga-cat.sh:1]: Do we really need this script?

Comment (stv0g) [reply src/villas-fpga-cat.sh:1]: Also the name _cat_ is misleading in my opinion, as its not doing what the Unix `cat` command is doing (con**cat**enating)

File: lib/fpgaHelper.cpp [hunk 9]
```diff
@@ -48,6 +49,64 @@ using namespace villas;
...
+	} else {
+		int port = std::stoi(str);
+
+		if (port > 7 || port < 0)
+			throw std::runtime_error("Invalid port number");
+
+		return aurora_channels[port];
+	}
+}
+// parses a string lik "1->2" or "1<->stdout" and configures the crossbar
```
Comment (stv0g) [inline lib/fpgaHelper.cpp:67]: Missing new line and wrong capitalization of comment.

File: lib/card.cpp [hunk 10]
```diff
@@ -215,6 +218,29 @@ std::shared_ptr<ip::Core> PCIeCard::lookupIp(const ip::IpIdentifier &id) const
...
 
+bool PCIeCard::unmapMemoryBlock(const MemoryBlock &block)
+{
+	if (memoryBlocksMapped.find(block.getAddrSpaceId()) == memoryBlocksMapped.end()) {
+		throw std::runtime_error("Block " + std::to_string(block.getAddrSpaceId()) + " is not mapped but was requested to be unmapped.");
+	}
+
+	auto &mm = MemoryManager::get();
+
+	// Unmap all memory blocks
```
Comment (stv0g) [inline lib/card.cpp:229]: From the code it looks like that this is not unmapping **all** blocks.

File: src/villas-fpga-cat.sh (hunk 8 above)
Comment (n-eiling) [reply src/villas-fpga-cat.sh:1]: this was already in the repo before and I want to avoid removing something without replacement.

**Comment by n-eiling** [comment]
I will not fix the minor style issues. Will be fixed by a clang-format task once this is available.

**Pull request merged**